idea/src/org/jetbrains/kotlin/idea/inspections/AnnotationTargetExpressionInspection.kt

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2010-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.idea.inspections
+
+import com.intellij.codeInspection.LocalQuickFix
+import com.intellij.codeInspection.ProblemDescriptor
+import com.intellij.codeInspection.ProblemHighlightType
+import com.intellij.codeInspection.ProblemsHolder
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.util.TextRange
+import com.intellij.psi.PsiElementVisitor
+import org.jetbrains.kotlin.builtins.KotlinBuiltIns
+import org.jetbrains.kotlin.idea.caches.resolve.resolveToCall
+import org.jetbrains.kotlin.psi.*
+import org.jetbrains.kotlin.psi.psiUtil.endOffset
+import org.jetbrains.kotlin.psi.psiUtil.startOffset
+import org.jetbrains.kotlin.resolve.descriptorUtil.fqNameSafe
+
+class AnnotationTargetExpressionInspection : AbstractKotlinInspection() {
+
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor {
+        return classVisitor(fun(element: KtClass) {
+            if (!element.isAnnotation()) return
+            val target = element.targetAnnotationArgument()?.asElement() ?: return
+            if (element.retentionAnnotation() == null) return
+
+            val start = element.startOffset
+            val highlightRange = TextRange(target.startOffset - start, target.endOffset - start)
+            holder.registerProblem(
+                holder.manager.createProblemDescriptor(
+                    element,
+                    highlightRange,
+                    "annotation will not be stored",
+                    ProblemHighlightType.GENERIC_ERROR_OR_WARNING,
+                    isOnTheFly,
+                    ChangeAnnotationRetentionToSourceFix()
+                )
+            )
+        })
+    }
+}
+
+private class ChangeAnnotationRetentionToSourceFix : LocalQuickFix {
+
+    private val source = AnnotationRetention.SOURCE.name
+
+    override fun getName() = "Change @Retention to $source"
+
+    override fun getFamilyName() = name
+
+    override fun applyFix(project: Project, descriptor: ProblemDescriptor) {
+        val element = descriptor.psiElement as? KtClass ?: return
+        val retention = element.retentionAnnotation() ?: return
+
+        val annotationTarget = "${KotlinBuiltIns.FQ_NAMES.annotationRetention.shortName().asString()}.$source"
+        val psiFactory = KtPsiFactory(element)
+        if (retention.valueArgumentList == null) {
+            retention.add(psiFactory.createCallArguments("($annotationTarget)"))
```

Comment:
Please insert the qualified name and then call ShortenReferences.
You code won't work with something like this:

```kotlin
class AnnotationRetention // unrelated class in the same package

@Retention()
@Target(AnnotationTarget.EXPRESSION)
annotation class Ann
```